File: frontend/src/interfaces/certificate.ts
import type { Post } from "./post";

export interface Certificatemanage {
    ID: number;
    title_th: string;
    title_en: string;
    detail: string;
    organizer: string;
    // picture_url: string;
    picture_participation?: string;
    picture_winner?: string;
    // signature_1_url: string;
    // signature_2_url: string;
    // date: string;
    type: string;
    post_id?: number;
    post?: Post;



}

export interface CreateCertificateRequest {

    title_th: string;
    title_en: string;
    detail: string;
    organizer: string;
    // picture_url: string;
    picture_participation?: string;
    picture_winner?: string;
    // signature_1_url: string;
    // signature_2_url: string;
    // date: string;
    type: string;

    user_id?: number;
    result_id?: number;
    registration_id?: number;
    post_id?: number;
    post?: Post;
}

export interface UpdateCertificateRequest {
    ID: number;
    title_th: string;
    title_en: string;
    detail: string;
    organizer: string;
    // picture_url: string;
    // signature_1_url: string;
    // signature_2_url: string;
    // date: string;
    type: string;
    post_id?: number;
    post?: Post;
}

export interface DeleteCertificateRequest {
    ID: number;
}

export interface GetCertificateResponse extends Certificatemanage { }

File: backend/entity/certificate.go
package entity

import (
	"github.com/asaskevich/govalidator"
	"gorm.io/gorm"
)

type Certificate struct {
	gorm.Model
	TitleTH       string `gorm:"not null" valid:"required~TitleTH is required,maxstringlength(200)~TitleTH is too long" json:"title_th"`
	TitleEN       string `json:"title_en"`
	Detail        string `json:"detail"`
	Organizer     string `json:"organizer"`
	PictureParticipation string `json:"picture_participation"`
	PictureWinner        string `json:"picture_winner"`
	Type          string `gorm:"not null" valid:"required~Type is required" json:"type"`

	UserID         *uint  `gorm:"not null" valid:"required~User ID is required" json:"user_id"`
	User           *User  `gorm:"foreignKey:UserID" json:"user,omitempty"`
	ResultID       *uint   `gorm:"" json:"result_id"`
	RegistrationID *uint   `gorm:"" json:"registration_id"`
	PostID         uint   `gorm:"not null" valid:"required~Post ID is required" json:"post_id"`
}

func ValidateCertificate(c Certificate) (bool, error) {
	return govalidator.ValidateStruct(c)
} 